server/rest_slicer_cli.py

```diff
@@ -826,14 +829,163 @@ def genRESTEndPointsForSlicerCLIsInDocker(info, restResource, dockerImages):
+        if id is None:
+            id = pullDockerImage(dimg)
+        if id is None:
+            raise ValueError("The image %s does not exist locally or "
+                             "on the default registry" % dimg)
+        # get CLI list
+        cliListSpec = localDockerImageCLIList(dimg)
+        if cliListSpec is None:
+            raise ValueError("Could not get the cli list for the img %s most "
+                             "likely the docker image entrypoint does "
+                             "not accept the argument --list_cli" % dimg)
+        # pprint.pprint(cliListSpec)
+
+        cliListSpec = json.loads(cliListSpec)
 
+        # Add REST end-point for each CLI
+        for cliRelPath in cliListSpec.keys():
+            cliXML = localDockerImageclixml(dimg, cliRelPath)
+            if cliXML is None:
+                raise ValueError("Could not get the cli for the img %s cli %s"
+                                 " most likely the docker image entrypoint does"
+                                 " not accept the argument %s --xml" %
+                                 (dimg, cliRelPath, cliRelPath))
+            # create a POST REST route that runs the CLI
+            try:
+
+                # print cliXML
+                cliRunHandler = genHandlerToRunDockerCLI(dimg,
+                                                         cliRelPath,
+                                                         cliXML,
+                                                         restResource)
+            except Exception as e:
+                print "Failed to create REST endpoints for %s: %s" % (
+                    cliRelPath, e)
+                print "cli handler"
+                continue
+
+            cliSuffix = os.path.normpath(cliRelPath).replace(os.sep, '_')
+
+            cliRunHandlerName = 'run_' + cliSuffix
+            setattr(restResource, cliRunHandlerName, cliRunHandler)
+            restResource.route('POST',
+                               (cliRelPath, 'run'),
+                               getattr(restResource, cliRunHandlerName))
+
+            # create GET REST route that returns the xml of the CLI
+            try:
+                cliGetXMLSpecHandler = genHandlerToGetDockerCLIXmlSpec(
+                    cliRelPath, cliXML, restResource)
+
+            except Exception as e:
+                print "Failed to create REST endpoints for %s: %s" % (
+                    cliRelPath, e)
+                exc_type, exc_obj, exc_tb = sys.exc_info()
+                fname = os.path.split(exc_tb.tb_frame.f_code.co_filename)[1]
+                print(exc_type, fname, exc_tb.tb_lineno)
+                continue
+
+            cliGetXMLSpecHandlerName = 'get_xml_' + cliSuffix
+            setattr(restResource,
+                    cliGetXMLSpecHandlerName,
+                    cliGetXMLSpecHandler)
+            restResource.route('GET',
+                               (cliRelPath, 'xmlspec',),
```

Comment:
Again just to keep in mind (no need to change now), to get around name conflicts across docker images, maybe we can prefix this route with a massaged version (maybe use hash if there is no generic way to massage) of the docker image string